### PR TITLE
Add RTEMS TCP/UDP Support

### PR DESCRIPTION
**Pull request by roncapat**
With this PR, the DDS Client (and subsequently micro-ros) can build and run on RTEMS 5.
RTEMS is a Real Time Executive in use in multiple research and space exploration projects (satellites, rovers...).

This is the issue where I kept track of the progress. 
https://github.com/micro-ROS/micro_ros_setup/issues/397
Today I managed to run a 1:1 copy of the ping-pong example with 50 Hz ping rate.

RTEMS support is enabled by defining `UCLIENT_PLATFORM_RTEMS_BSD_NET`. 

Both TCP and UDP transports have been tested.

### CODE REVIEW

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/466/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/480/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/469/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/483/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/471/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/485/)

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/472/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/486/)

**Comment by pablogs9** [comment]
Thanks @roncapat, let's review it and we can merge it in develop branch

File: src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        uint8_t* errcode)
+{
+    size_t rv = 0;
+
+    struct timeval tv;
+    tv.tv_sec = timeout / 1000;
+    tv.tv_usec = (timeout % 1000) * 1000;
+
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wsign-conversion"
```
Comment (Acuadros95) [inline src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c:109]: Why is this necessary?

File: src/c/profile/discovery/transport/udp_transport_datagram_rtems_bsd_net.c [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        const uint8_t* bytes,
+        char* ip)
+{
+    uint32_t addr;
+    addr = (uint32_t)(*bytes + (*(bytes + 1) << 8) + (*(bytes + 2) << 16) + (*(bytes + 3) << 24));
+    struct in_addr saddr;
+    saddr.s_addr = addr;
+    char* res = inet_ntoa(saddr);
+    strncpy(ip, res, 16); // 16 == MAX_IPv4_LEN
+
```
Comment (pablogs9) [inline src/c/profile/discovery/transport/udp_transport_datagram_rtems_bsd_net.c:125]: Unnecessary new lines

File: src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        platform->remote_addr.sin_port = htons(iport);
+        platform->remote_addr.sin_addr.s_addr = inet_addr(ip);
+
+        /* Poll setup. */
+        #pragma GCC diagnostic push
+        #pragma GCC diagnostic ignored "-Wsign-conversion"
+        FD_ZERO(&platform->select_fd);
+        FD_SET(platform->fd, &platform->select_fd);
+        #pragma GCC diagnostic pop
+        rv = true;
```
Comment (Acuadros95) [inline src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c:51]:
```suggestion
```

Redundant line with 
https://github.com/eProsima/Micro-XRCE-DDS-Client/blob/e31c74728820f9897f868c9ddcb5ca3a663448ce/src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c#L57

File: src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c (hunk 1 above)
Comment (roncapat) [reply src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c:109]: To prevent a GCC warning on those lines. It is a common and known "problem" with those directives ([one of google results fo example](https://www.linuxquestions.org/questions/programming-9/impossible-to-use-gcc-with-wconversion-and-standard-socket-macros-841935/)) but if you search there are lots.

If you want I can remove the preprocessor directives, but then some compiler may fire warnings that don't seem to have a proper solution outside suppression.

File: CMakeLists.txt [hunk 4]
```diff
@@ -159,6 +161,8 @@ endif()
 # Transport sources.
 set(_transport_src)
 
+set(UCLIENT_PLATFORM_RTEMS_BSD_NET ON)
```
[the remark this answers is not in the document]
Comment (roncapat) [reply CMakeLists.txt:164]: Ahh yes! Sorry, I missed this line while cleaning.

File: src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c (hunk 1 above)
Comment (pablogs9) [reply src/c/profile/transport/ip/tcp/tcp_transport_rtems_bsd_net.c:108]: I'm ok with this. pragmas will not interfere with other compilers.

**Comment by richiprosima** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Linux/477/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/Micro-XRCE-DDS-Client%20Manual%20Windows/493/)

**Review by Acuadros95 — APPROVED** [review]

**Review by pablogs9 — APPROVED** [review]

**Comment by roncapat** [comment]
Thank you so much! :100:

**Comment by pablogs9** [comment]
@roncapat we are really interested in your use case. Will you be interested in talking about the purpose / use cases / projects where RTEMS is going to be used along with micro-ROS in our monthly ROS 2 Embedded Working Group?

**Comment by roncapat** [comment]
@pablogs9 of course I am. 

Note that currently, this is a low-priority and personal contribution, but is meant to pave the way to internal usage (at Thales Alenia Space Italy, where I currently work) in some robotic testbenches. Let me know more about the meeting at roncapat@gmail.com (so that I can then reply with my corporate mail).

File: src/c/profile/discovery/transport/udp_transport_datagram_rtems_bsd_net.c [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        int timeout)
+{
+    bool rv = false;
+
+    struct timeval tv;
+    tv.tv_sec = timeout / 1000;
+    tv.tv_usec = (timeout % 1000) * 1000;
+
+
+    int32_t poll_rv = select(transport->fd + 1, &transport->select_fd, NULL, NULL, &tv);
```
Comment (roncapat) [inline src/c/profile/discovery/transport/udp_transport_datagram_rtems_bsd_net.c:95]: Bug: The same FD_ZERO FD_SET treatment before select() should be done here. Will post a new PR soon.

Comment (pablogs9) [reply src/c/profile/discovery/transport/udp_transport_datagram_rtems_bsd_net.c:95]: Cool, thanks. We will handle it in a new PR.

Comment (Acuadros95) [reply src/c/profile/discovery/transport/udp_transport_datagram_rtems_bsd_net.c:95]: Yes! We missed that the set object is modified.

We also use a select approach on the multi-serial agent, where it is enough to copy the set on a local variable, check here: [link](https://github.com/eProsima/Micro-XRCE-DDS-Agent/blob/develop/src/cpp/transport/serial/MultiSerialAgentLinux.cpp#L82) (`read_fds` is the equivalent to your `transport->select_fd`).

With this approach you avoid modifications on external data.

Comment (roncapat) [reply src/c/profile/discovery/transport/udp_transport_datagram_rtems_bsd_net.c:95]: I like it, cleaner! I will also then update the other two usages of select() like this.
